mariamihaelafainarea/assignment3/Assignment3

```diff
@@ -0,0 +1,192 @@
+        String currentString = start.toString();
+        int[] dx = {-1,-1,-1,0,0,1,1,1};
+        int[] dy = {-1,0,1,-1,1,-1,0,1};
+        for (int i = 0; i < 8; i++) {
+            int newX = x + dx[i];
+            int newY = y + dy[i];
+            if (newX >= 0 && newX < n) {
+                if (newY >= 0 && newY < m) {
+                    if(visited[newX][newY] == 0) {
+                        visited[newX][newY] = 1;
+                        StringBuilder hopa = start.append(grid[newX][newY]);
+                        if (dictionary.isPrefix(hopa.toString())){
+                            if (dictionary.isWord(hopa.toString())) {
+                                theList.add(hopa.toString());
+                            }
+                            DFSonGrid(theList, hopa, newX, newY,visited);
+                        }
+                        //restart the state after a deep search ! 
+                        start.deleteCharAt(hopa.length()-1);
+                        visited[newX][newY] = 0;
+                    }
+                }
+            }
+        }
+        
+    }
+    class MyDictionary{
+        //A dictionary based on a trie ! 
+        //It has a better complexity than
+        //a hashmap !
+        private class TrieNode{
+            private HashMap<Character,TrieNode> children;
+            private boolean endOfWord;
+            //endOfWord is like.. here in this node
+            //a word ends even if it is not a leaf !
+            public TrieNode() {
+                children = new HashMap<>();
+                endOfWord = false;
+            }
+            
+        }
+        private final TrieNode root;
+        
+        public void insert(String word) {
+            //the idea is very simple
+            //you start from root
+            //you go deep down in the trie 
+            //with the current node
+            //if theChildren (the character that need to be added)
+            //exists you only pass it and go deep down
+            TrieNode current = root;
+            for(int i = 0; i < word.length();i++) {
+                TrieNode theChildren = current.children.get(word.charAt(i));
+                if (theChildren == null) {  
+                    theChildren = new TrieNode();
+                    current.children.put(word.charAt(i), theChildren);
+                }
+                current = theChildren;
+            }
+            //after you add the last node and move your
+            //current there
+            //you mark it as endOfWord
+            current.endOfWord = true;
+            
+        }
+        public boolean isWordHelper(TrieNode current,String word) {
+            if (word.equals("")) {
+                return current.endOfWord;
+            }
+            TrieNode theChildren = current.children.get(word.charAt(0));
+            if (theChildren == null) {
+                return false;
+            }
+            return isWordHelper(theChildren, word.substring(1));
```

Comment:
Just a thought: maybe having recursive calls with a reference to the same string and two indexes (or maybe even one) would be better for performance. The method substring makes a copy of a string every time so basically you end up with an expensive stack of recursive calls where every one has its own instance of the string. And that string could potentially be really long

Reply:
done